Foundation/Network/HostName.hsc

```diff
@@ -0,0 +1,166 @@
+  deriving (Show, Eq, Ord, Typeable, Hashable)
+instance IsString HostName where
+    fromString = HostName . fromString
+
+-- | HostName's Info
+data HostNameInfo address_type = HostNameInfo
+    { officialName :: !HostName
+        -- ^ official names
+    , aliases      :: ![HostName]
+        -- ^ known aliases
+    , addresses    :: ![address_type]
+        -- ^ known addresses
+    } deriving (Show, Eq, Ord, Typeable)
+
+-- | HostName errors
+data HostNameError
+    = HostNotFound !HostName
+        -- ^ the given HostMame was not found
+    | NoAssociatedData !HostName
+        -- ^ there is not associated info/data to the given HostName
+        --
+        -- i.e. : no IPv4 info? This might mean you should try IPv6 ?
+    | FatalError
+        -- ^ getHostNameInfo uses *C* a binding to get the `HostNameInfo`
+        --
+        -- a fatal error is linked to the underlying *C* function and is not
+        -- recoverable.
+    | UnknownError !CInt
+        -- ^ Unknown Error, `CInt` is the associated error code.
+        --
+        -- see man gethostbyname for more information
+  deriving (Show,Eq,Typeable)
+
+instance Exception HostNameError
+
+-- | get `HostName` info:
+--
+-- retrieve the official name, the aliases and the addresses associated to this
+-- hostname.
+--
+-- For cross-platform compatibility purpose, this function is using a *C* non
+-- re-entrant function `gethostbyname2`. This function is using a `MVar ()` to
+-- avoid a race condition and should be safe to use.
+--
+getHostNameInfo :: (Show address_type, Eq address_type, Address address_type)
+                => HostName
+                -> IO (HostNameInfo address_type)
+getHostNameInfo = getHostNameInfo_ Proxy
+
+globalMutex :: MVar ()
+globalMutex = unsafePerformIO (newMVar ())
+
+-- | like `getHostNameInfo` but takes a `Proxy` to help with the type checker.
+getHostNameInfo_ :: (Show address_type, Eq address_type, Address address_type)
+                 => Proxy address_type
+                 -> HostName
+                 -> IO (HostNameInfo address_type)
+getHostNameInfo_ p h@(HostName hn) =
+    withMVar globalMutex $ \_ ->
+    withCString (toList hn) $ \cname -> do
+        ptr <- loop $ c_gethostbyname2 cname (familyCode p)
+
+        cptr <- peek $ castPtr (offname_ptr ptr)
+        on <- getCString cptr
+
+        alptr <- peek $ castPtr (aliases_ptr ptr)
+        alptr' <- peek alptr
+        as <- getAliases alptr'
+
+        ty <- peek $ castPtr (addtype_ptr ptr)
+        len <- peek $ castPtr (length_addr ptr)
+        lptr <- peek $ castPtr (addr_list ptr)
+        lptr' <- peek lptr
+        addrs <- getAddresses p ty len lptr'
+        return $ HostNameInfo (HostName on) as addrs
+  where
+    loop f = do
+        ptr <- f
+        if ptr /= nullPtr
+            then return ptr
+            else do
+                err <- c_get_h_errno
+                case err of
+                    _ | err == (#const HOST_NOT_FOUND) -> throwIO $ HostNotFound h
```

Comment:
yep, didn't really noticed the h. it really need a better bindings, and moving to System.Bindings too.